File: src/apps/api/MedVault.API/Features/Shared/Domain/Gender.cs
namespace MedVault.API.Features.Shared.Domain;

public enum Gender
{
    Male,
    Female,
    Other,
    PreferNotToSay
}

public static class GenderExtensions
{
    public static Gender? ToGender(this string? value)
    {
        if (string.IsNullOrWhiteSpace(value))
        {
            return null;
        }

        return value.Trim() switch
        {
            nameof(Gender.Male) => Gender.Male,
            nameof(Gender.Female) => Gender.Female,
            nameof(Gender.Other) => Gender.Other,
            nameof(Gender.PreferNotToSay) => Gender.PreferNotToSay,
            "Prefer not to say" => Gender.PreferNotToSay,
            "prefer-not-to-say" => Gender.PreferNotToSay,
            _ => Enum.TryParse<Gender>(value, ignoreCase: true, out var gender)
                ? gender
                : null,
        };
    }
}

File: src/apps/mobile/medvault/lib/models/gender.dart
enum Gender { male, female, other, preferNotToSay }

extension GenderApiValue on Gender {
  String get apiValue {
    switch (this) {
      case Gender.male:
        return 'Male';
      case Gender.female:
        return 'Female';
      case Gender.other:
        return 'Other';
      case Gender.preferNotToSay:
        return 'PreferNotToSay';
    }
  }
}

Gender? genderFromApiValue(String? value) {
  switch (value) {
    case 'Male':
    case 'male':
      return Gender.male;
    case 'Female':
    case 'female':
      return Gender.female;
    case 'Other':
    case 'other':
      return Gender.other;
    case 'PreferNotToSay':
    case 'preferNotToSay':
    case 'Prefer not to say':
    case 'prefer-not-to-say':
      return Gender.preferNotToSay;
    default:
      return null;
  }
}
